backend/server.js

```diff
@@ -195,30 +247,49 @@ sbp('sbp/selectors/register', {
       // This is done in a queue to handle several simultaneous requests
       // reading and writing to the same key
       await appendToIndexFactory(resourcesKey)(resourceID)
+      // Done as a persistent action to return quickly. If one of the owners
+      // up the chain has many resources, the operation could take a while.
+      sbp('chelonia.persistentActions/enqueue', ['backend/server/addToIndirectResourcesIndex', resourceID])
     })
   },
+  'backend/server/addToIndirectResourcesIndex': async function (resourceID: string) {
+    const ownerID = await sbp('chelonia.db/get', `_private_owner_${resourceID}`)
+    let indirectOwnerID = ownerID
+    // If the owner of the owner doesn't exist, there are no indirect resources.
+    while ((indirectOwnerID = await sbp('chelonia.db/get', `_private_owner_${indirectOwnerID}`))) {
+      await appendToIndexFactory(`_private_indirectResources_${indirectOwnerID}`)(resourceID)
+    }
+  },
+  'backend/server/removeFromIndirectResourcesIndex': async function (resourceID: string) {
+    const ownerID = await sbp('chelonia.db/get', `_private_owner_${resourceID}`)
+    const resources = await sbp('chelonia.db/get', `_private_resources_${resourceID}`)
+    const indirectResources = resources ? await sbp('chelonia.db/get', `_private_indirectResources_${resourceID}`) : undefined
+    const allSubresources = [
+      resourceID,
+      ...(resources ? resources.split('\x00') : []),
+      ...(indirectResources ? indirectResources.split('\x00') : [])
+    ]
+    let indirectOwnerID = ownerID
+    while ((indirectOwnerID = await sbp('chelonia.db/get', `_private_owner_${indirectOwnerID}`))) {
+      await removeFromIndexFactory(`_private_indirectResources_${indirectOwnerID}`)(allSubresources)
+    }
+  },
   'backend/server/registerBillableEntity': appendToIndexFactory('_private_billable_entities'),
-  'backend/server/updateSize': async function (resourceID: string, size: number) {
+  'backend/server/updateSize': function (resourceID: string, size: number) {
     const sizeKey = `_private_size_${resourceID}`
-    if (!Number.isSafeInteger(size)) {
-      throw new TypeError(`Invalid given size ${size} for ${resourceID}`)
-    }
-    // Use a queue to ensure atomic updates
-    await sbp('okTurtles.eventQueue/queueEvent', sizeKey, async () => {
-      // Size is stored as a decimal value
-      const existingSize = parseInt(await sbp('chelonia.db/get', sizeKey, 10) ?? '0')
-      if (!(existingSize >= 0)) {
-        throw new TypeError(`Invalid stored size ${existingSize} for ${resourceID}`)
-      }
-      await sbp('chelonia.db/set', sizeKey, (existingSize + size).toString(10))
-    })
+    return updateSize(resourceID, sizeKey, size)
+  },
+  'backend/server/updateContractFilesTotalSize': function (resourceID: string, size: number) {
+    const sizeKey = `_private_contractFilesTotalSize_${resourceID}`
+    return updateSize(resourceID, sizeKey, size)
```

Comment:
I'm noticing that this calls `updateSize`, which calls `'worker/updateSizeSideEffects'`, but `'worker/updateSizeSideEffects'` only processes updates if `sizeKey.startsWith('_private_size_')`...

So why call `updateSize` here instead of calling `updateSize_`?

That way you could also remove the `if` check for `_private_size_` in `'worker/updateSizeSideEffects'`